projects/cm_mcu/LocalTasks.c

```diff
@@ -335,20 +339,121 @@ struct MonitorI2CTaskArgs_t clockr0a_args = {
     .stack_size = 4096U,
 };
 
-bool isEnabledFF(int ff)
+void setFFmask(uint32_t ff_combined_present)
 {
-  // firefly config stored in on-board EEPROM
-  static bool configured = false;
 
-  static uint32_t ff_config;
-  if (!configured) {
-    ff_config = read_eeprom_single(EEPROM_ID_FF_ADDR);
-    configured = true;
-  }
-  if (!((1 << ff) & ff_config))
+  log_info(LOG_SERVICE, "Setting a bit mask of enabled Fireflys to 1 \r\n");
+
+  uint32_t data = (~ff_combined_present) & 0xFFFFFU; // the bit value for an FF mask is an inverted bit value of the PRESENT signals
+
+  ff_USER_mask = read_eeprom_single(EEPROM_ID_FF_ADDR);
+  ff_PRESENT_mask = data;
+  uint64_t block = EEPROMBlockFromAddr(ADDR_FF);
+
+  uint64_t unlock = EPRMMessage((uint64_t)EPRM_UNLOCK_BLOCK, block, PASS);
+  xQueueSendToBack(xEPRMQueue_in, &unlock, portMAX_DELAY);
+
+  uint64_t message = EPRMMessage((uint64_t)EPRM_WRITE_SINGLE, ADDR_FF, data);
+  xQueueSendToBack(xEPRMQueue_in, &message, portMAX_DELAY);
+
+  uint64_t lock = EPRMMessage((uint64_t)EPRM_LOCK_BLOCK, block << 32, 0);
+  xQueueSendToBack(xEPRMQueue_in, &lock, portMAX_DELAY);
+
+  return;
+}
+
+void readFFpresent()
+{
+
+  while (xSemaphoreTake(i2c4_sem, (TickType_t)10) == pdFALSE)
+    ;
+#ifdef REV1
+  uint32_t present_FFLDAQ_F1, present_FFL12_F1, present_FFLDAQ_0X20_F2, present_FFL12_0X20_F2, present_FFLDAQ_0X21_F2, present_FFL12_0X21_F2;
+#elif defined(REV2)
+  uint32_t present_FFLDAQ_F1, present_FFL12_F1, present_FFLDAQ_F2, present_FFL12_F2;
+#endif
+
+#ifdef REV1
+  // to port 7
+  apollo_i2c_ctl_w(4, 0x70, 1, 0x80);
+  apollo_i2c_ctl_reg_r(4, 0x20, 1, 0x01, 1, &present_FFL12_F1);
+  // to port 6
+  apollo_i2c_ctl_w(4, 0x71, 1, 0x40);
+  apollo_i2c_ctl_reg_r(4, 0x21, 1, 0x00, 1, &present_FFLDAQ_F1);
+#elif defined(REV2)
+  // to port 7
+  apollo_i2c_ctl_w(4, 0x70, 1, 0x80);
+  apollo_i2c_ctl_reg_r(4, 0x20, 1, 0x01, 1, &present_FFL12_F1);
+  // to port 6
+  apollo_i2c_ctl_w(4, 0x71, 1, 0x40);
+  apollo_i2c_ctl_reg_r(4, 0x21, 1, 0x00, 1, &present_FFLDAQ_F1);
+#endif
+
+  xSemaphoreGive(i2c4_sem); // if we have a semaphore, give it
+
+  while (xSemaphoreTake(i2c3_sem, (TickType_t)10) == pdFALSE)
+    ;
+#ifdef REV1
+  // to port 0
+  apollo_i2c_ctl_w(3, 0x72, 1, 0x01);
+  apollo_i2c_ctl_reg_r(3, 0x20, 1, 0x01, 1, &present_FFL12_0X20_F2);
+  apollo_i2c_ctl_reg_r(3, 0x21, 1, 0x01, 1, &present_FFL12_0X21_F2);
+  // to port 1
+  apollo_i2c_ctl_w(3, 0x72, 1, 0x04);
+  apollo_i2c_ctl_reg_r(3, 0x20, 1, 0x01, 1, &present_FFLDAQ_0X20_F2);
+  apollo_i2c_ctl_reg_r(3, 0x21, 1, 0x01, 1, &present_FFLDAQ_0X21_F2);
```

Comment:
If I'm not mistaken, the lines at 398-400 are repeated in lines 402-404 but just stored in different variables? Looking at the schematic I'm confused what the difference is between the FFL12 and the FFLDAQ variables.

Reply:
ok wait now I see that the last argument is different on the mux. my bad.